### ISSUE TITLE
[BUG] FrameLimiter Sleep strategy sleeps too long

### ISSUE THREAD

**Issue by AkkaWakka** [amethyst/amethyst#2083]
I think there is an error in the FrameLimiter implementation logic for the `Sleep` strategy, unless I'm misunderstanding something.

The docs say (emphasis mine):
> Sleep will call `thread::sleep` with a **duration of 0 milliseconds** until the frame duration has passed.

In the code, the `Sleep` strategy is called with a zero duration:
```rust
const ZERO: Duration = Duration::from_millis(0);
....
Sleep => self.do_sleep(ZERO)
```

Working through the `do_sleep` function (comments mine, durations expressed in milis):
```rust
fn do_sleep(&self, stop_on_remaining: Duration) {    // stop_on_remaining = 0 milis
    let frame_duration = self.frame_duration - stop_on_remaining;    // assume 16ms frame, frame_duration = 16 - 0 = 16 milis
    loop {
        let elapsed = Instant::now() - self.last_call;    // assume last_call 4ms, elapsed = 4 milis
        if elapsed >= frame_duration {    // if 4 > = 16 is false
            break;
        } else {    // so we end up here
            sleep(frame_duration - elapsed);    // sleep(16 - 4) | sleep(12)
        }
    }
}
```

I think the error here is that the `stop_on_remaining` parameter is being used for two different things:
1) For the `SleepAndYield` strategy, setting the time before the next frame that we must stop sleeping and start yielding.  This - I think - works.
2) For the `Sleep` strategy, setting the time we want to sleep for, each time we call sleep.

I see two possible solutions:
a) add another parameter to the `do_sleep` method.  Something like `max_sleep_length`.  This might be useful for the `SleepAndYield` strategy too since it would remove the possibility of very long sleeps (and thread unresponsiveness)
b) use a separate method or loop for the `Sleep` strategy.

I don't know this code well, and I'm finding the names a little obtuse, so please tell me if I've got this wrong.

Thanks

**Comment by zaethan** [comment]
I _think_ the code is fine. The documentation seems to be outdated though. If you look at the git blame of the frame limiter file, you can see that the documentation was written two years ago. However, the do_sleep function was changed just 11 months ago.
Before that change, the do_sleep function would indeed call thread::sleep repeatedly with 0 ms until the remaining time was less than the stop_on_remaining parameter.

Now however, the function calculates how long it needs to sleep (so that stop_on_remaining milliseconds are left) and then sleeps once for that amount of time. After the thread has finished sleeping `elapsed >= frame_duration` should be true.
I would say this actually works fine for both methods, since the sleep strategy is just a special case of sleepandyield (0 ms remaining).

**Comment by caitp** [comment]
FYI, I believe this is fixed by #2156.  It doesn't look like there was any code change needed, and the original filing was based on reading the code/docs, and not on a problem observed at runtime.

I avoided including `Fixes #2156` in the commit message to avoid spamming this bug :)

**Comment by All8Up** [comment]
You might want to revisit the code even though it is "logically" correct.  The likely issue is that sleep is notoriously inaccurate on most OS's and tends to overshoot the requested time.  I.e. you want to sleep for 12ms, it doesn't wake up for 20.  In order to solve this, the common method is to sleep for 'half' the desired time in a loop until you get "close enough".  This is much better than consistently overshooting, which is the common behavior.  Some OS's (see remarks: https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep), it can be on modulus of 10-16ms (see remarks: https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-gettickcount).

I don't know if Chrome (or MS) fixed how they set high precision mode but it used to work as a "fix" for any program with a poorly programmed frame rate limiter because high precision mode timing is/was a global setting in Windows and Chrome always set it on startup.  So, just opening Chrome and putting it in the background used to fix many games.  Of course leaving high precision mode on is a bad idea as Chrome found out because it burns through laptop batteries since the Os is now servicing things at a greatly increased rate than normal.

Just my $0.02 since I've run into this where someone did the above type of code assuming sleep was accurate.

**Comment by Kintar** [comment]
@All8Up : Are you still planning to look into this, or can the ticket be considered up for grabs?

**Comment by All8Up** [comment]
> @All8Up : Are you still planning to look into this, or can the ticket be considered up for grabs?

Work is keeping me quite busy at the moment.  So it is unlikely I could say when I'd be able to look at this in more detail.  All said and done, I don't suggest waiting on me right now.  :)

**Comment by borup3** [comment]
I have opened a PR on this issue

**Comment by W4RH4WK** [comment]
@borup3 Have you consider sleeping for shorter periods of time as well as busy waiting instead of yielding?

As reference, we are using a frame limiter that (targeting 60 FPS) sleeps for 2 ms intervals and busy waits if the remaining time is smaller than 2 ms. Frame times turn out to be pretty accurate.

Intervals for engaging sleep periods as well as the threshold for starting the busy wait loop can be adjusted per platform / use case.

**Comment by All8Up** [comment]
> 
> 
> @borup3 Have you consider sleeping for shorter periods of time as well as busy waiting instead of yielding?
> 
> As reference, we are using a frame limiter that sleeps for the predicted frame time minus some margin (2 ms when targeting 60 FPS). A busy loop follows until the reach the desired frame time. Frame times turn out to be pretty accurate.

Unfortunately it's not this easy, especially on Windows.  The problem is that the timer in the scheduler has a fixed tick rate which is used to trigger thread wakes.  Depending on the version of Windows it is usually set to 10 or 15 ms.  So, obviously if you sleep 10 and the resolution is set to 15, you will have issues.  Worse though is the granularity is fixed time ticking such that if you call sleep(10) just 1 ms after the last tick, the next tick will show your timer with 1ms left so it won't wake the thread, it will instead wait till the next tick, i.e. 19ms from the time you called sleep.  Now add on top of this that folks like to play with timeBeginPeriod which can change the granularity globally it impacts everything in the system so you have to deal with it randomly changing.  (Folks that use timeBeginPeriod; sleep(xx); timeEndPeriod; should be taken out and flogged, really screws up the system. :) )

All said and done, the proper solution would go find the granularity of the thread scheduler (applies to all OS's, not just Windows) and keep checking it once in a while in case someone fiddles with it (only Windows I believe).  With that in hand you can use sleep for anything > granularity*2 expecting to have to busy wait for any amount of time between 0 and granularity.   Writing the busy wait is a bit of a trick though since on Intel you really want to be using _mm_pause.  The reason for that is to prevent the CPU from thinking it is doing real work and as such never throttling down when possible.  Other CPU's have different instructions of course, but they are usually easy to find.  Finally the busy wait shouldn't target the entire time left, it should basically do wait 1/2 remaining, then 1/2 the remains after that etc until within 1ms or so.  Otherwise various thread hitches can throw off the timing also.

Hope that answers why this isn't quite the trivial fix..  :)

**Comment by W4RH4WK** [comment]
> Unfortunately it's not this easy, especially on Windows. The problem is that the timer in the scheduler has a fixed tick rate which is used to trigger thread wakes. Depending on the version of Windows it is usually set to 10 or 15 ms. So, obviously if you sleep 10 and the resolution is set to 15, you will have issues. Worse though is the granularity is fixed time ticking such that if you call sleep(10) just 1 ms after the last tick, the next tick will show your timer with 1ms left so it won't wake the thread, it will instead wait till the next tick, i.e. 19ms from the time you called sleep. Now add on top of this that folks like to play with timeBeginPeriod which can change the granularity globally it impacts everything in the system so you have to deal with it randomly changing. (Folks that use timeBeginPeriod; sleep(xx); timeEndPeriod; should be taken out and flogged, really screws up the system. :) )
>
> All said and done, the proper solution would go find the granularity of the thread scheduler (applies to all OS's, not just Windows) and keep checking it once in a while in case someone fiddles with it (only Windows I believe). With that in hand you can use sleep for anything > granularity*2 expecting to have to busy wait for any amount of time between 0 and granularity. 

I assumed that `timeBeginPeriod(1)` is called during platform initialization like most interactive applications would do on Windows. While it changes the granularity globally, it seems to honor the smallest value demanded by running applications ([see](https://github.com/CRYTEK/CRYENGINE/blob/6c4f4df4a7a092300d630f8f89d2ebda39183c36/Code/CryEngine/CrySystem/PlatformOS/PlatformOS_PC.cpp#L668)). I'd assume that other platforms provide a similar mechanism to guarantee a reasonable granularity.

> Writing the busy wait is a bit of a trick though since on Intel you really want to be using _mm_pause. The reason for that is to prevent the CPU from thinking it is doing real work and as such never throttling down when possible. Other CPU's have different instructions of course, but they are usually easy to find.

I definitely get the idea of using `_mm_pause`, just not sure how big the impact is with a frame limiter. 

> Finally the busy wait shouldn't target the entire time left, it should basically do wait 1/2 remaining, then 1/2 the remains after that etc until within 1ms or so. Otherwise various thread hitches can throw off the timing also.

With `_mm_pause` you cannot provide a specific time, can you? It pretty much works as a *nop* for a couple of cycles ([see](https://stackoverflow.com/a/5833693)). So you'd probably call a bunch of them and poll the current time stamp in a loop, adjusting the amount of `_mm_pause` calls accordingly, right?

Do we have some specific use case or benchmark we could utilize to determine whether the gain of this method is significant?

**Comment by All8Up** [comment]
> I assumed that `timeBeginPeriod(1)` is called during platform initialization like most interactive applications would do on Windows. While it changes the granularity globally, it seems to honor the smallest value demanded by running applications ([see](https://github.com/CRYTEK/CRYENGINE/blob/6c4f4df4a7a092300d630f8f89d2ebda39183c36/Code/CryEngine/CrySystem/PlatformOS/PlatformOS_PC.cpp#L668)). I'd assume that other platforms provide a similar mechanism to guarantee a reasonable granularity.

So, having done this for titles in the past I have two specific reasons to suggest not using timeBeginPeriod at all.  One, it cranks everything in the OS, i.e, it is servicing interrupts that much faster, rescaning the thread control blocks (and process blocks) etc 10-15 times more often etc.  This is bad in many ways but mostly because it is blowing out the cache more regularly and memory is the #1 performance limiter on any system that goes parallel on more than 6ish cores.  Silly example, I tested with a 3990x when I first saw this thread because I was thinking maybe MS had fixed some of this....  23% higher kernel times on my massively parallel simulation engine, nope not fixed.  (Overall it caused a near 30% slowdown.)  The second reason to keep away from this is related but more towards mobile gaming, a laptop battery is going to drain considerably faster if you muck with time period, really suggest not doing it.

And please, in the future, never ever link anything from CryEngine, I never want to see/work on that code base ever again!  :)  CryEngine is #2 on my worst code base list, only beat out #1 The Sims...

>>    Finally the busy wait shouldn't target the entire time left, it should basically do wait 1/2 remaining, then 1/2 the remains after that etc until within 1ms or so. Otherwise various thread hitches can throw off the timing also.
> With _mm_pause you cannot provide a specific time, can you? It pretty much works as a nop for a couple of cycles (see). So you'd probably call a bunch of them and poll the current time stamp in a loop, adjusting the amount of _mm_pause calls accordingly, right?

Basically when you use _mm_pause you just put it in a loop for a certain amount of iterations then recheck the HPC to see how more to iterate.  It's basically the opposite of an exponential backoff solution for threading work.  Of course there is another gotcha here, in their infinite wisdom, Intel made the duration of _mm_pause fixed on "most" CPU's but then went and changed it on Xeon.  So, you have to detect the CPU to know how many iterations to use.  Unfortunately the difference is significant such that if you don't do this it cripples one or the other CPU.

> Do we have some specific use case or benchmark we could utilize to determine whether the gain of this method is significant?
Unfortunately none of the times I have implemented this or variations have source available.  I've not seen many games do a reasonable job of this overall.  The solution in CryEngine works, but that engine doesn't scale to more than 4 or 5 threads worth a damn before Mr Amdahl's law nut punches the performance, so they don't notice the massive impact using timeBeginPeriod is having on the overall machine.

As a note, with holidays and such coming up and having mostly finished a rush contract, I might be able to take some time and do some of this work..  We'll see.

**Comment by borup3** [comment]
There's a curious level of pedanticism here. Do I sense a pet peeve?

> @borup3 Have you consider sleeping for shorter periods of time as well as busy waiting instead of yielding?
> 
> As reference, we are using a frame limiter that sleeps for the predicted frame time minus some margin (2 ms when targeting 60 FPS). A busy loop follows until the reach the desired frame time. Frame times turn out to be pretty accurate.

@W4RH4WK Frame times were appropriately accurate with yielding that there was no need to be power hungry. If you need _that_ precise a frame limiter then I would go for a more bespoke solution than the current FrameLimiter altogether as an alternative solution, e.g. like you've done or such as All8Up's elusive strategy being on offer. For a large majority of scenarios, the simple solution will solve the problem sufficiently.

**Comment by All8Up** [comment]
> There's a curious level of pedanticism here. Do I sense a pet peeve?

Only in that I've had to fix it several times because the timeBeginPeriod method gets added and all of a sudden all the KPI's start going nuts.  CPU times go way up, performance drops etc..  So, yup, pet peeve is probably accurate though it is also just that over 30 years of making games this "trivial" thing always seems to turn out to be not so trivial mostly on Windows for all the various reasons I've mentioned.  Additionally, I've had to fight with this in a threading scheduler for our simulation engine and in the case of 128 threads on the 3990x, doing it "perfectly" is pretty much a requirement.

**Comment by W4RH4WK** [comment]
> All said and done, the proper solution would go find the granularity of the thread scheduler (applies to all OS's, not just Windows) and keep checking it once in a while in case someone fiddles with it (only Windows I believe).

I've just checked this (using [Timer Tool](https://vvvv.org/contribution/windows-system-timer-tool)):

On my main machine, freshly booted, balanced power profile, *Current* is stuck at ~ 0.75 ms. I have a few applications in startup, but none that would obviously require such a low value. My laptop, freshly booted, starts with a value of around 15 ms (default) and quickly changes between ~ 1 ms, ~ 4 ms, ~ 6 ms. There I don't have any power hungry applications / services in startup.

It seems to me that, unless you have a very clean system, scheduler granularity is either already quite low, or jumping around. I get the idea that people want to minimize maintaining a low granularity to improve power efficiency, but even the [docs](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep#remarks) recommend against that.

> With that in hand you can use sleep for anything > granularity*2 expecting to have to busy wait for any amount of time between 0 and granularity.

The first part assumes that we can rely on the queried granularity to remain constant until we engage the sleep. Not sure how valid this assumption is, or how big the impact of violating it is. Just wanted to highlight it.

Busy waiting would then be the obvious choice, but there goes our power efficiency, doesn't it?

> The second reason to keep away from this is related but more towards mobile gaming, a laptop battery is going to drain considerably faster if you muck with time period, really suggest not doing it.

I am starting to doubt that we can have a one-size-fits-all solution here. We might distinguish between either more power efficiency (mobile) or more accurate frame times (desktop). Of course there is also the cache issue which is also relevant for desktop. But as described above, that ship has probably sailed on most users desktop systems.

> So, you have to detect the CPU to know how many iterations to use. Unfortunately the difference is significant such that if you don't do this it cripples one or the other CPU.

With such a low-level performance optimization, I'd expect to have various special cases for different platforms, CPUs, OS versions, etc. While I respect the urge to *get things right*, I imagine it's pretty hard to pull the trigger one this without a representative benchmark. The implementation would introduce additional complexity and should be well worth the trade-off.

> Frame times were appropriately accurate with yielding that there was no need to be power hungry.

👍

> As a note, with holidays and such coming up and having mostly finished a rush contract, I might be able to take some time and do some of this work.. We'll see.

Looking forward to it. Let me know if there is anything I could help out with, find me on Discord (same username).

- - -

> And please, in the future, never ever link anything from CryEngine.

Initially I wanted to link to Unreal, but not everyone has direct access to view the source. I believe there is some registration you have to complete beforehand. 

> CryEngine is nr2 on my worst code base list, only beat out nr1 The Sims...

I can imagine 😉, at least there are comments and they are not in Japanese 😜.

**Comment by All8Up** [comment]
> > With that in hand you can use sleep for anything > granularity*2 expecting to have to busy wait for any amount of time between 0 and granularity.

> Busy waiting would then be the obvious choice, but there goes our power efficiency, doesn't it?

That's why you use _mm_pause, the instruction is purpose built in order to idle the CPU for a fixed period without raised power consumption.  This is how you write exponential backoff for atomic operations (i.e. user space mutex/condvar etc) without causing major energy consumption issues.

>>    So, you have to detect the CPU to know how many iterations to use. Unfortunately the difference is significant such that if you don't do this it cripples one or the other CPU.

> With such a low-level performance optimization, I'd expect to have various special cases for different platforms, CPUs, OS versions, etc. While I respect the urge to get things right, I imagine it's pretty hard to pull the trigger one this without a representative benchmark. The implementation would introduce additional complexity and should be well worth the trade-off.

Yup, this is why one of the first things I write is a set of KPI's (Key Performance Indicators, i.e. fps for a single triangle and then scene, time to path from one side of a map to the other, memory usage, CPU times etc etc) that run automatically at least on a daily basis, preferably after every commit if you can get away with it.  Especially when you start muckin with threaded code it is incredibly easy to do something trivial which has major impacts.  Windows is a major culprit in this due to some API's that are more than 30 years old which don't explain clearly that they are replaced with versions that don't do stupid things like system wide pauses.

** Having said all this, I'm actually taking a look at the results I got on the 3990x, something doesn't add up given information I ran across last night when my interest was piqued.  I.e. "supposedly" Win8 did actually attempt to fix this mess to a degree and moved to a tickless scheduler which is supposed to have adaptive resolution.  But this is not what I saw on Win 10 Pro running the 64 core beast,  Figuring out why that was showing terrible results might make this easier.  For my purposes on that machine I have to have the closest to perfect result possible, this is a problem for all multicore solutions as latency for work triggering is basically multiplied by the number of threads, i.e. 128 ms worth of latency possibly for 1 ms slop, not good. **

**Comment by W4RH4WK** [comment]
Just stumbled over [this blog post](https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/). It's rather recent and my provide a tiny bit more insight.

**Comment by All8Up** [comment]
> Just stumbled over [this blog post](https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/). It's rather recent and may provide a tiny bit more insight.

Thanks, that is very interesting and does explain a few odd results I got but ignored thinking I was just getting bad data.  Unfortunately the numbers on the 3990x are considerably different than any I have seen so far which has me stumped at the moment.  Primarily the massive performance drop when I insert a begin/end period call, not to mention the latency between execution stages spikes into the 10+ ms range which destroys performance even further.  I'm writing some code (in Rust so maybe it will turn out helpful here) to test some guesses, we'll see.  I have tomorrow off and nothing to do anyway so ...  a hackin I go, oh a hackin I go ....

**Comment by All8Up** [comment]
> With such a low-level performance optimization, I'd expect to have various special cases for different platforms, CPUs, OS versions, etc. While I respect the urge to get things right, I imagine it's pretty hard to pull the trigger one this without a representative benchmark. The implementation would introduce additional complexity and should be well worth the trade-off.

So, I put together a very stupid starting point for testing: https://github.com/All8Up/test_limiter (Be kind, it's quick and dirty at the moment..)  The numbers are interesting to say the least.  On the 3950x, not the 3990x yet, Windows is pretty much as bad as last time I did this.  Running a simple 5 second per implementation test you end up at 7.5 secs for a pure sleep based solution, 5.5 is for a timeBeginPeriod based solution and 4.9 for a quick and dirty timeBeginPeriod with exponential decay.  So, total errors end up around 50%, 10% and 2%.  2% is fairly reasonable for a game engine, especially if you err on the low side rather than the high side since things like present with v-sync and such will be much happier that way.

To make this future proofed, it will need to do actual work across all cores and measure user/kernel time utilization to see what the implementations are doing to the OS and thread scheduling itself.  Especially at higher core counts such as a 3990x I'm still concerned about the overhead introduced by cranking the scheduler, but we'll see with some numbers hopefully shortly.

**Comment by saveriomiroddi** [comment]
Hi there!

This is a very interesting subject, so I'll chip in with a few informations. Feel free to ignore any, as I'm not an expert :smile:

## spin-sleep crate

I'm personally not very convinced by this library. I've raised my concerns on an issue I've opened [today](https://github.com/alexheretic/spin-sleep/issues/7). Ultimately, it doesn't do significantly more than the existing code:

1. it uses timeBeginPeriod() on Windows, exactly the way described above as `(Folks that use timeBeginPeriod; sleep(xx); timeEndPeriod; should be taken out and flogged, really screws up the system. :) )`
2. on the other systems it uses simple thread yielding (which is approximately what Amethyst does)

## pause intrinsic

For general information, curiosity and lulz, Rust implements the pause intrinsic: [`std::hint::spin_loop()`](https://doc.rust-lang.org/std/hint/fn.spin_loop.html), which is portable and O/S-independent (but I'm not implying that busy-loop is the way to go).

## general considerations

> With such a low-level performance optimization, I'd expect to have various special cases for different platforms, CPUs, OS versions, etc. While I respect the urge to get things right, I imagine it's pretty hard to pull the trigger one this without a representative benchmark. The implementation would introduce additional complexity and should be well worth the trade-off.

I support this consideration, and I'd go as far as waiting (=closing) until a user brings up the issue with a real-world use-case (this issue itself was opened because of a presumed issue with the code, not because of somebody experiencing the problem).

As a matter of fact, even if a solution is found for Windows, other operating systems may require entirely different solutions (I can't trivially find the Amethyst supported O/Ss, but I guess it's the usual Linux/Windows/OS X). I'd find problematic if the the solution was implemented for a single O/S, considering the portability of the project and the Rust culture in general.

AFAIK scheduling has no better solution in an absolute sense, and any other option is just a different position on the spectrum precision/consumption (with the two ends being thread sleeping <> busy wait).

> Do we have some specific use case or benchmark we could utilize to determine whether the gain of this method is significant?

Indeed; to put the problem in another way, before talking about implementation details, I think a use case should be found :)

I've quickly checked the official examples, and for example, `arc_ball_camera` and/or `fly_camera` could be used as reference, with programmatic input; there's actually plenty of stuff [here](https://community.amethyst.rs/t/games-made-with-amethyst/134/8) that fits as well.

> but that engine doesn't scale to more than 4 or 5 threads worth a damn

The devil is in the details :)

4 or 5 threads should not be underestimated. Last year I was doing some experimentations on lock-free concurrency for a project of mine, and with my great horror, after 4 threads, performance degraded significantly.

This is anecdata, but the point is that considering highly concurrent use cases is (currently) far fetched IMO.

## conclusion

All in all though, I think that producing test cases that prove or disprove the necessity of improving the existing code, could be provided fairly easily (within the limits of small-scale testing), since I think the issue depends on the context (O/S and other applications running) rather than the program logic itself (if needed, artificial slowdowns can be easily added).

I'll work on it; I should provide the use cases in two solar weeks. I'm a Rust advanced beginner, and not a game developer :)

**Comment by All8Up** [comment]
> ## pause intrinsic
> 
> For general information, curiosity and lulz, Rust implements the pause intrinsic: [`std::hint::spin_loop()`](https://doc.rust-lang.org/std/hint/fn.spin_loop.html), which is portable and O/S-independent (but I'm not implying that busy-loop is the way to go).

This is the proper solution if you want really high accuracy.  But, keep in mind that when you use this you are typically implementing an exponential backoff solution such that you pause 16 times the first loop, 32 times the second ect till you get to some max.  Or in this case you start at some big number and reduce as you get closer to the target time.  The problem is that those numbers are CPU specific.  This is why there is justifiable hesitation to go down this path, you have to add CPU detection and/or timing mechanics to adjust this to the specific machine.  It's not horrible, just annoying.

> ## general considerations
> 
> > With such a low-level performance optimization, I'd expect to have various special cases for different platforms, CPUs, OS versions, etc. While I respect the urge to get things right, I imagine it's pretty hard to pull the trigger one this without a representative benchmark. The implementation would introduce additional complexity and should be well worth the trade-off.
> 
> I support this consideration, and I'd go as far as waiting (=closing) until a user brings up the issue with a real-world use-case (this issue itself was opened because of a presumed issue with the code, not because of somebody experiencing the problem).

I wouldn't argue with this though I will point out the prior experience here that if you don't get it under about 5% accuracy (which the quick and dirty example I made achieves) you are likely to have just enough of a jitter to miss v-sync occasionally (or have rapidly changing rates with g-sync) which tends to make the overall FPS just a little inconsistent and rather annoying to an end user.  This is generally subtle but just notable enough that folks have the feeling something is not quite right and small details on screen tend to jitter enough to be annoying.

> > Do we have some specific use case or benchmark we could utilize to determine whether the gain of this method is significant?
> 
> Indeed; to put the problem in another way, before talking about implementation details, I think a use case should be found :)

This is why I mentioned KPI's, personally I have two favorites as a starter for most engine work:
1.  A couple simple pixel, or really small entities, that just bounce around the screen.  If everything is working correctly it looks perfectly smooth.  But, if the latency jitter is high you will easily see inconsistent motion.  This is a very visual thing of course and it's likely to be better and worse on different machines.  Getting it consistent across machines is the goal.  The key idea here is that small, medium speed items tend to show the latency issue fairly clearly.  With a full scene, with a lot of things going on tends to hide the impact but the human eye still detects it and makes it feel wrong, but it's hard to point out what the actual issue is.  So, this simple little testbed is helpful to point it out easily.
2. A simple flocking simulation.  These are excellent for a number of reasons in that they are effectively the most brutal memory access and CPU stress test you can perform if you don't put in the simple "cheat" optimizations.  I.e. the neighbor search should use proper per entity accurate searching, not the common optimization using flow fields.  Done without cheats, a 3950x + 3200Mhz memory should be able to handle about 50k entities at 90Hz, that is absolutely maxing out memory bandwidth, several execution pipes (preferably SIMD pipes) and other items of course but also requires massive parallelism that most engines just can't touch.

> 4 or 5 threads should not be underestimated. Last year I was doing some experimentations on lock-free concurrency for a project of mine, and with my great horror, after 4 threads, performance degraded significantly.

Threading is a fun, and difficult, problem.  After quite a few years in the field, it is my personal opinion that if you have to use lock-free containers as the core of your performance, you've already failed.  As minor optimizations, great, but if you can architect around them that's your primary goal.  Unfortunately game engines tend to focus on work stealing queues which I tend to borrow a Star Wars Meme when discussing: "It's a trap".  The problem is that in order to fill many threads you have to break down the work into smaller pieces, unfortunately that exponentially increases the contention on the queue and Amdahl's law starts kicking you squarely in the nether regions.

> This is anecdata, but the point is that considering highly concurrent use cases is (currently) far fetched IMO.
> ## conclusion
> 
> All in all though, I think that producing test cases that prove or disprove the necessity of improving the existing code, could be provided fairly easily (within the limits of small-scale testing), since I think the issue depends on the context (O/S and other applications running) rather than the program logic itself (if needed, artificial slowdowns can be easily added).
> 
> I'll work on it; I should provide the use cases in two solar weeks. I'm a Rust advanced beginner, and not a game developer :)

Keep in mind that CPU utilization (i.e. entity counts) is not the only, or even the most important, reason to worry about concurrency.  If you go look up the voltage versus frequency of most CPU's you see a very interesting curve.  Generalizing a bunch here but you will find that the most effective voltage vs frequency is around 25% of max frequency.  Call that the '1.0' sweet spot.  When you get to 100% frequency on the core, most CPU's are sitting at 5 to 8+ times that 1.0 voltage level.  So, as an example of why concurrency is good even for a simple game, my last contract was to fix the battery life of a popular simple mobile game.  I applied some concurrency to the main loop and particle systems and quadrupled the average battery life simply due to the CPU being able to scale down the frequencies into the sweet spot.  Literally no optimizations, no fancy lazy evaluations etc, just apply a bit of concurrency.  It took longer to explain why that would work better than their initial intentions than it did to actually implement it.

Another use case, which is where I do most of my work, is involving cloud based servers.  There are a number of different pricing models of course and depending on model you either want to minimize utilization of any single core as with the above, or you want to absolutely burn every possible core without wasting anything at all.  With the second goal, the primary limiting factor is how many game servers you can consolidate onto a single instance, if the threading is not effective (and these sorts of problems are directly related) you will be limited.  Simply "optimizing" more code, which is the common approach, is a diminishing returns approach, better concurrency across the board can give you exponential gains for similar work investment.

So, while testing with high concurrency is generally what you do, the benefit is not purely how many entities you can drive, it's how overall efficient it can be done for the above and other less important reasons.

*** As much as I'd like to take the little testbed I put together and go further with it, trying to get funding for one company and doing contract work with the other has kept me incredibly busy.  Hopefully things will calm down a bit if we close the funding.

**Comment by saveriomiroddi** [comment]
Small update: the `spin-sleep` crate author replied with benchmarks that are.. [gulp!](https://github.com/alexheretic/spin-sleep/issues/1#issuecomment-372304712).

Based on his benchmarks, in short, the thread yielding strategy is, within a negligible margin, as precise as the pause intrinsic strategy.

In best case scenario, the current strategies don't need significant changes (possibly, the only one being the `timeBeginPeriod()` usage, which requires a judgment call from somebody experienced).

I'll produce data to make an informed decision (in particular, the benchmark above seems to be a microbenchmark, which needs to be carefully reviewed), while also trying to understand how thread yielding can be so precise on Linux; the performance is so good that I don't exclude that the compiler (and/or the kernel) are doing special optimizations.

**Comment by saveriomiroddi** [comment]
So!

Time's up, so here there are my findings.

## Setups and limitations

There are three sets of results, based on two projects (Pong tutorial 6 and Evoli) and two setups (a Linux Ryzen 3950x/GT 710 and a Windows 10 Core i5-6300U/Intel integrated).

Evoli crashed immediately on Windows 10. I've tried a third project, Dwarf seeks fortune, but couldn't compile it due to an unclear (to me) feature problem.

Additionally, on the Linux machine, the CPU occupation was really high. Even pong used something like 140% of CPU as minimum.

Nonetheless, I think that the results are indicative. Most importantly though, I've designed the change so that anybody interested just need to change a `Cargo.toml` line, `cargo run` the projects, and get a convenient CSV with the results.

## Running and design

In order to run the modified (pseudo-"profiled" version), change the `amethyst` dependency to point to the fork:

```toml
[dependencies]
amethyst = { git = "https://github.com/saveriomiroddi/amethyst-dev", branch = "REFERENCE/profiled_frame_limiter/0.15.3", features = ["vulkan"] }
```

When the project is run, it will run as usual, while dumping under `framelimiter_data` a CSV with useful data (`frame_duration_ms,frame_work_ms,expected_sleep_ms,actual_sleep_ms,skew_ms`).

I've created two branches, one for 0.15.3, and the other for 0.15.0. The HEAD commit of each branch includes the change itself.

It's easier to just check it out rather than reading a description :)

## Findings

I'm attaching to this comment the result sets I've collected (I had to append the `.txt` extension due to GitHub limitations):

- [pong_tutorial_06_linux_sleep.csv.txt](https://github.com/amethyst/amethyst/files/6216518/pong_tutorial_06_linux_sleep.csv.txt)
- [pong_tutorial_06_linux_yield.csv.txt](https://github.com/amethyst/amethyst/files/6216519/pong_tutorial_06_linux_yield.csv.txt)
- [pong_tutorial_06_windows_sleep.csv.txt](https://github.com/amethyst/amethyst/files/6216520/pong_tutorial_06_windows_sleep.csv.txt)
- [pong_tutorial_06_windows_yield.csv.txt](https://github.com/amethyst/amethyst/files/6216521/pong_tutorial_06_windows_yield.csv.txt)
- [evolution-island_linux_sleep.csv.txt](https://github.com/amethyst/amethyst/files/6216516/evolution-island_linux_sleep.csv.txt)
- [evolution-island_linux_yield.csv.txt](https://github.com/amethyst/amethyst/files/6216517/evolution-island_linux_yield.csv.txt)

Now, with the giant disclaimer that I don't know if Amethyst is supposed to take so much CPU on Linux (I can't exclude it's a known issue, or similar), and that this is a very small number of setups, my summary, based on Pong 6, is:

- Sleep strategy
  - Linux was remarkly punctual when using the sleep strategy; it drifted of around 50/60 µs on each cycle, which is remarkable - with results like this, it could be a valid and efficient solution
  - Windows was poor, instead; it drifted systematically of 2 to 15 ms; this is clearly a non-viable solution
    - The CPU usage was very low, though; pong used ~2%
- Yield strategy
  - Linux was almost exact here; the drift was almost negligible (<= 1 µs)
    - The CPU went from sky-high to moon-high - from 140% to something like 250%
  - On Windows, the drift went down to the same drift as linux (~ 1 µs)
    - The CPU went higher, however, to a sufficiently low level - ~17%

The results on Evoli, which could run on Linux only, the results were the same a Pong 6.

## Further tests

Based on a Discord discussion with Jazarro, I may be able to compile also Dwarf seeks fortune, however, it would be interesting to have the tests run on other setups. If useful, I should be able to artificially limit the cores on my machine, but it'd still be best to run on a entirely different setups.

## Conclusions

Within limits & disclaimers etc.etc., based on the above, the conclusions are:

1. Linux: the sleep strategy could be a good default (but it should be tested on a machine with much fewer cores)
2. Windows: the yield strategy is necessary; on the other hand

On both cases, the drift is so small, that there isn't any necessity (unless use real-world cases show it is) to implement more advanced techniques.

## Notes

Both Evoli and Dwarf seeks fortune crashed when I hovered with the mouse over the window.

I can't exclude that there are bugs in my code; it's very simple though, if there is any, it should be easy to spot. As somebody said, "I have only proved it correct, not tried it"... and I didn't even prove it correct! :smile:

**Comment by saveriomiroddi** [comment]
Further notes: I have the sneaky suspicion that the yielding strategy is not significantly different than a standard spin loop with intrinsic - in order words, precise, but expensive; the spin-sleep crate results unfortunately doesn't include CPU occupation. It's easy to create an Amethyst/Pong fork with a SpinLoop strategy and verify (Windows would be best), but given the results above, it may be irrelevant (but still interesting :wink:).

**Comment by W4RH4WK** [comment]
Hey there @saveriomiroddi, thanks for the insight, great work.

I am following this conversation only loosely, but this part seems a bit suspicious.

> Windows was poor, instead; it drifted systematically of 2 to 15 ms

You probably know this already but just to be sure. Are you calling `timeBeginPeriod`? The span from 2 to 15 ms seems like some other application is continuously adjusting scheduler resolution globally and you are just experiencing whatever value it's been set to.

We discussed setting `timeBeginPeriod` previously. I'll also link the helpful blog post about non-global behavior in recent versions of Windows 10 again [here](https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/).

**Comment by All8Up** [comment]
> I am following this conversation only loosely, but this part seems a bit suspicious.
> 
> > Windows was poor, instead; it drifted systematically of 2 to 15 ms
> 
> You probably know this already but just to be sure. Are you calling `timeBeginPeriod`? The span from 2 to 15 ms seems like some other application is continuously adjusting scheduler resolution globally and you are just experiencing whatever value it's been set to.

Actually, that is to be expected if you don't use timeBeginPeriod.  With a typical default resolution of 10ish ms it's more surprising that it's even that accurate.  :D  Unfortunately Windows is a real pig in this area and tends to be ever changing as to what the best practice is.  But, I do believe that the little example I put together which uses timeBeginPeriod "once" at startup and then uses an exponential decay approach is still likely to be the best mix of accuracy without screwing up Windows in general.

And as a note, anything which drives the cpu utilization up too far (i.e. spin waits) is going to be undesirable.  It won't matter for a desktop but laptops and mobile devices will burn batteries.  Also, another target where it is undesirable, any client/server setup, a server will be using a limiter to tick at say 10-60 FPS, if the CPU is sitting high for no reason then auto instance scaling such as K8's will be unable to work correctly.  So, not a very good thing there.

Such a simple thing to do conceptually, such a crock to get correct.  :D

**Comment by W4RH4WK** [comment]
> Actually, that is to be expected if you don't use timeBeginPeriod.

Yes, that's why I am bringing it up again.

> anything which drives the cpu utilization up too far (i.e. spin waits) is going to be undesirable. It won't matter for a desktop but laptops and mobile devices will burn batteries.

We still don't have any data regarding the battery usage impact of `timeBeginPeriod` on a up-to-date system. Maybe it doesn't have any significant impact, in which case the sleep variant would be fine.

**Comment by saveriomiroddi** [comment]
> Hey there @saveriomiroddi, thanks for the insight, great work.
> 
> I am following this conversation only loosely, but this part seems a bit suspicious.
> 
> > Windows was poor, instead; it drifted systematically of 2 to 15 ms
> 
> You probably know this already but just to be sure. Are you calling `timeBeginPeriod`? The span from 2 to 15 ms seems like some other application is continuously adjusting scheduler resolution globally and you are just experiencing whatever value it's been set to.
> 
> We discussed setting `timeBeginPeriod` previously. I'll also link the helpful blog post about non-global behavior in recent versions of Windows 10 again [here](https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/).

Hi there!

So! This was essentially the first iteration. The target was to see on real projects the performance of the existing timing strategy - I was aware of `timeBeginPeriod`, but as first step, I wanted to be sure that tweaking is needed. 

Today I've compiled and timed Dwarf seeks fortune on Windows, and the results are substantially similar to the Pong 6 ones:

- sleep: CPU=~2/3%, skew=~0/15ms
- yield: CPU=~25%, skew=<1ms

Now, one note: I'm going to put Linux testing on hold. The CPU occupation is far too high, and I can't consider it stable and valid (again, I can't be certain whether it's my setup or Amethyst).

As next step, I'll add other two frame limiting strategies:

- the "tweaked" windows sleep
- the pause intrinsic

Having the datasets at hand will allow an informed decision. Until we don't have the full data, it's not possible to assume that a given strategy is absolutely better than another/the others (considering the two variables: CPU occupation and precision).

**Comment by saveriomiroddi** [comment]
I'm likely busy this week, so I'll follow up the week after.

I'll also look into implement CPU occupation monitoring, so that the CSVs will include the whole picture, but I can't guarantee that this will be possible (or straightforward; keeping the code simple may have higher priority).

**Comment by saveriomiroddi** [comment]
Hi there!

So, I've completed all the implementations and testing today.

I've implemented two new limiters, `SpinLoop` and `SleepTweaked`. The second is a sleep tweaked with `timeBeginPeriod`.

The implementation of everything is on the branch [mentioned above](https://github.com/saveriomiroddi/amethyst-dev/commits/REFERENCE/profiled_frame_limiter/0.15.3), which can also be trivially used, again, as mentioned above, in any project.

I'm experiencing another issue which may be a bug with Amethyst - the frame compution taking much a disproportionally high time when the frame rate is relatively high. See the description in [last commit of the branch](https://github.com/saveriomiroddi/amethyst-dev/commit/09ea85c9b595070e560f73c7368360e39faec26c).

Anyway, here there are the resulting CSVs:

- [pong_tutorial_06_sleep_tweaked_windows.csv.txt](https://github.com/amethyst/amethyst/files/6217501/pong_tutorial_06_sleep_tweaked_windows.csv.txt)
- [pong_tutorial_06_sleep_windows.csv.txt](https://github.com/amethyst/amethyst/files/6217502/pong_tutorial_06_sleep_windows.csv.txt)
- [pong_tutorial_06_spin_loop_windows.csv.txt](https://github.com/amethyst/amethyst/files/6217503/pong_tutorial_06_spin_loop_windows.csv.txt)
- [pong_tutorial_06_yield_windows.csv.txt](https://github.com/amethyst/amethyst/files/6217504/pong_tutorial_06_yield_windows.csv.txt)

Very roughly, the cpu loads (as reported by the Task Manager) are:

- sleep: 3%
- sleep_tweaked: 3%
- yield: 26%
- spin_loop: 28%

Note that the non-sleep strategies above are unfavorable in cases of simple games. In real games, where the game logic requires more computation, the CPU occupation of the wait strategy values will be proportionally inferior.  
Still, the general considerations apply.

## Caveats/notes

- In this run, sleep didn't have high skew as in yesterday's run. I have no idea why; given the "sleep tweaked" strategy, this fact shouldn't be relevant anymore.
- From the logs, I've noticed that the frame limiter (and profiler) are instantiated twice; I don't know if this is intended, and especially, if this affects the results (I suspect not). This could probably be verified by logging a message when the frame limiter is dropped.

## Conclusions

Based on all the data at hand:

- using the sleep tweaked (on Windows) seems to be a no-brainer; it resulted in a sub-ms skew, which I think is enough for the engine (as it is now)
- the yield strategy is almost as demanding as the spin loop
- the spin loop is, to no surprise, the most demanding, although only marginally more than the yield
- on Linux, with the mentioned big disclaimer, the standard sleep may be good enough, but it'd be best to have data from other setups (specifically, CPUs with few cores)

I have a power meter, so I'm going to check the power consumption; once I'm done with that, I think we have all the tools and data :smile:

**Comment by saveriomiroddi** [comment]
So!

This is the power data; they're readings from the wall socket; they're not meant to be exact, but indicative - 0.5W of difference are meant to be a generic "slightly more":

- no game running: 6W
- sleep: 9W
- sleep tweaked: 9.5W
- yield: 17W
- spin loop: 17W

I think that's the last of my data :smile: 

I can provide an implementation, if a design decision is taken (which may require, or not, additional data from other setups).

I'd personally just tweak the existing sleep, without creating any other strategy (or modifying existing ones), as the accuracy reached, and power consumption, are good enough.

**Comment by W4RH4WK** [comment]
> I'd personally just tweak the existing sleep, without creating any other strategy (or modifying existing ones), as the accuracy reached, and power consumption, are good enough.

I agree. A ~ 5% increase in power consumption seems like an acceptable tradeoff.

**Comment by saveriomiroddi** [comment]
Hey all, I've created the MVPR :wink: for this change. I didn't did it before, because I was hoping for other devs to test the strategy. I've seen at least another project making the same change (that is, using the `spin-sleep` crate), so I think this is essentially a fairly standard approach.
